rust/kernel/spi.rs

```diff
@@ -0,0 +1,180 @@
+    spi_driver: Option<bindings::spi_driver>,
+}
+
+impl DriverRegistration {
+    fn new(
+        this_module: &'static crate::ThisModule,
+        name: CStr<'static>,
+        probe: Option<SpiMethod>,
+        remove: Option<SpiMethod>,
+        shutdown: Option<SpiMethodVoid>,
+    ) -> Self {
+        DriverRegistration {
+            this_module,
+            name,
+            registered: false,
+            probe,
+            remove,
+            shutdown,
+            spi_driver: None,
+        }
+    }
+
+    // FIXME: Add documentation
+    pub fn new_pinned(
+        this_module: &'static crate::ThisModule,
+        name: CStr<'static>,
+        probe: Option<SpiMethod>,
+        remove: Option<SpiMethod>,
+        shutdown: Option<SpiMethodVoid>,
+    ) -> Result<Pin<Box<Self>>> {
+        let mut registration = Pin::from(Box::try_new(Self::new(
+            this_module,
+            name,
+            probe,
+            remove,
+            shutdown,
+        ))?);
+
+        registration.as_mut().register()?;
+
+        Ok(registration)
+    }
+
+    // FIXME: Add documentation
+    pub fn register(self: Pin<&mut Self>) -> Result {
+        let mut spi_driver = bindings::spi_driver::default();
+        spi_driver.driver.name = self.name.as_ptr() as *const c_types::c_char;
+        spi_driver.probe = self.probe;
+        spi_driver.remove = self.remove;
+        spi_driver.shutdown = self.shutdown;
+
+        let this = unsafe { self.get_unchecked_mut() };
+        if this.registered {
+            return Err(Error::EINVAL);
+        }
+
+        this.spi_driver = Some(spi_driver);
+
+        let res = unsafe { bindings::__spi_register_driver(this.this_module.0, &mut spi_driver) };
+
+        match res {
+            0 => {
+                this.registered = true;
+                Ok(())
+            }
+            _ => Err(Error::from_kernel_errno(res)),
+        }
+    }
+}
+
+impl Drop for DriverRegistration {
+    fn drop(&mut self) {
+        unsafe { bindings::driver_unregister(&mut self.spi_driver.unwrap().driver) }
+        // FIXME: No unwrap? But it's safe?
+    }
+}
+
+// FIXME: Fix SAFETY documentation
+
+// SAFETY: The only method is `register()`, which requires a (pinned) mutable `Registration`, so it
+// is safe to pass `&Registration` to multiple threads because it offers no interior mutability.
+unsafe impl Sync for DriverRegistration {}
+
+// SAFETY: The only method is `register()`, which requires a (pinned) mutable `Registration`, so it
+// is safe to pass `&Registration` to multiple threads because it offers no interior mutability.
+unsafe impl Send for DriverRegistration {}
+
+type SpiMethod = unsafe extern "C" fn(*mut bindings::spi_device) -> c_types::c_int;
+type SpiMethodVoid = unsafe extern "C" fn(*mut bindings::spi_device) -> ();
+
+#[macro_export]
+macro_rules! spi_method {
+    (fn $method_name:ident (mut $device_name:ident : SpiDevice) -> Result $block:block) => {
```

Comment:
> So you take the Rust `probe()` method from the driver, transform it into an `extern "C"` using a macro, store it inside the `Registration`, then provide that to `extern "C" spi_register()`, is that correct?

Yep, that's exactly it.

> There are various reasons why, in Rust, template functions/classes and Traits are preferred over macros. I leave the explanation to real experts such as @alex, @ojeda or @wedsonaf.

Yeah, we are already seeing the limits of our macro, as calling it on a function without the correct keywords produces a really hard to understand error. We did that because we had no idea on how to do it properly, without having the user define functions using C types directly. 
  
> `file_operations.rs` provides an idiomatic example of how to do this: drivers implement a `pub trait` provided by the core, and pass that to the registration function. This has the added benefit that drivers can just leave out the trait functions they're not interested in. Instead of having to pass `None`, `None`, ...
> 
> Perhaps follow that example for `spi` ?

This is definitely a way better solution. We did the macro because we were having a hard time understanding the entirety of `FileOperations` and needed a quick way to abstract the usage of those methods. We will look into reworking this similarly to what has been achieved in `file_operations.rs`